src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,245 @@
+    altHoldState->targetAltitude = getCurrentAltitude(altHoldState);
+    altHoldState->smoothedAltitude = 0.01f;
+}
+
+void altHoldInit(altHoldState_s* altHoldState)
+{
+    altHoldState->altHoldEnabled = false;
+    altHoldState->throttleFactor = 0.0f;
+    altHoldState->velocityEstimationAccel = 0.0f;
+    altHoldReset(altHoldState);
+}
+
+void altHoldProcessTransitions(altHoldState_s* altHoldState) {
+    bool newAltHoldEnabled = FLIGHT_MODE(ALTHOLD_MODE);
+
+    if (FLIGHT_MODE(GPS_RESCUE_MODE) | failsafeIsActive()) {
+        newAltHoldEnabled = false;
+    }
+
+    if (newAltHoldEnabled && !altHoldState->altHoldEnabled)
+    {
+        altHoldReset(altHoldState);
+    }
+    if (!newAltHoldEnabled && altHoldState->altHoldEnabled) {
+        altHoldState->exitTime = millis();
+    }
+    altHoldState->altHoldEnabled = newAltHoldEnabled;
+
+    uint32_t currTime = millis();
+
+    if (newAltHoldEnabled) {
+        uint32_t timeSinceEnter = currTime - altHoldState->enterTime;
+        if (timeSinceEnter < ALTHOLD_ENTER_PERIOD) {
+            float delta = (float)timeSinceEnter / ALTHOLD_ENTER_PERIOD;
+            altHoldState->throttleFactor = MAX(delta, altHoldState->throttleFactor);
+            return;
+        }
+        altHoldState->throttleFactor = 1.0f;
+        return;
+    }
+
+    if (altHoldState->exitTime == 0) {
+        altHoldState->throttleFactor = 0.0f;
+        return;
+    }
+
+    uint32_t timeSinceExit = currTime - altHoldState->exitTime;
+    if (timeSinceExit < ALTHOLD_MAX_EXIT_PERIOD) {
+        float delta = (float)timeSinceExit / ALTHOLD_MAX_EXIT_PERIOD;
+        altHoldState->throttleFactor = MIN(altHoldState->throttleFactor, 1.0f - delta);
+        return;
+    }
+
+    altHoldState->throttleFactor = 0.0f;
+}
+
+void altHoldUpdate(altHoldState_s* altHoldState)
+{
+    altHoldProcessTransitions(altHoldState);
+
+    float timeInterval = 1.0f / ALTHOLD_TASK_PERIOD;
+
+    float measuredAltitude = getCurrentAltitude(altHoldState);
+
+    t_fp_vector accelerationVector = {{
+        acc.accADC[X],
+        acc.accADC[Y],
+        acc.accADC[Z]
+    }};
+
+    imuTransformVectorBodyToEarth(&accelerationVector);
+
+    float measuredAccel = 9.8f * (accelerationVector.V.Z - acc.dev.acc_1G) / acc.dev.acc_1G;
+
+    DEBUG_SET(DEBUG_ALTHOLD, 0, (int16_t)(measuredAccel * 100.0f));
+
+    altHoldState->velocityEstimationAccel += measuredAccel * timeInterval;
+    altHoldState->velocityEstimationAccel *= 0.999f;
+
+    float currentVelocityEstimationAccel = altHoldState->velocityEstimationAccel - altHoldState->startVelocityEstimationAccel;
+    DEBUG_SET(DEBUG_ALTHOLD, 1, (int16_t)(100.0f * currentVelocityEstimationAccel));
+
+    altHoldState->measuredAltitude = measuredAltitude;
+    altHoldState->measuredAccel = measuredAccel;
+
+    float velocityTarget = simplePidCalculate(&altHoldState->altPid, timeInterval, altHoldState->targetAltitude, altHoldState->measuredAltitude);
```

Comment:
Shouldn't this parameter be called 'altitude target' since the P and I values are both targeting altitude error?

Reply:
The naming here is because it calculates a target velocity for a velocity controller. Probably could be changed.